File: constraint-prototype/test/red_pandas/resolution_test.clj
(ns red-pandas.resolution-test
  (:require
   [red-pandas.resolution :as rs]
   [red-pandas.TypeError]
   [red-pandas.core]
   [clojure.test :as t])
  (:import
   [red_pandas TypeError]
   [red_pandas.core Fail]))

(def rules [[(rs/pred :father :john :mary)]
            [(rs/pred :father :john :alice)]
            [(rs/pred :mother :susan :mary)]
            [(rs/pred :mother :susan :alice)]
            (rs/fresh [x y]
              [(rs/pred :parent x y)
               (rs/pred :father x y)])
            (rs/fresh [x y]
              [(rs/pred :parent x y)
               (rs/pred :mother x y)])])
;;(def query (fresh [a] (pred :father a :mary)))

;; (resolve-pretty query rules :parallel? false)


(t/deftest test-resolution
  (t/testing "father(X, Y)"
    (t/is (= [(rs/pred :father :john :mary)
              (rs/pred :father :john :alice)]
             (rs/resolve-pretty (rs/fresh [a b] (rs/pred :father a b)) rules))))

  (t/testing "mother(X, Y)"
    (t/is (= [(rs/pred :mother :susan :mary)
              (rs/pred :mother :susan :alice)]
             (rs/resolve-pretty (rs/fresh [a b] (rs/pred :mother a b)) rules))))
  
  (t/testing "mother(X, :mary)"
    (t/is (= [(rs/pred :mother :susan :mary)]
             (rs/resolve-pretty (rs/fresh [a] (rs/pred :mother a :mary)) rules))))
  (t/testing "fail"
    (t/is (thrown-with-msg? TypeError #"test"
                            (rs/resolve-pretty (Fail. "test") rules)))))

File: constraint-prototype/src/test/scala/ResolutionTest.scala
package red_pandas

val rules: List[Rule] = List(
  List(Predicate("father", List("John", "Mary"))),
  List(Predicate("father", List("John", "Alice"))),
  List(Predicate("mother", List("Susan", "Alice"))),
  List(Predicate("mother", List("Susan", "Mary"))), {
    val x = new Variable("X")
    val y = new Variable("Y")
    List(
      Predicate("parent", List(x, y)),
      Predicate("father", List(x, y))
    )
  }, {
    val x = new Variable("X")
    val y = new Variable("Y")
    List(
      Predicate("parent", List(x, y)),
      Predicate("mother", List(x, y))
    )
  }
)

class ResolutionTest extends munit.FunSuite {
  test("father(X, Y)") {
    val expected = List(
      List(Predicate("father", List("John", "Alice"))),
      List(Predicate("father", List("John", "Mary")))
    )
    val goal =
      List(Predicate("father", List(new Variable("X"), new Variable("Y"))))
    assertEquals(resolve_pretty(goal, rules), expected)
  }

  test("mother(X, Y)") {
    val expected = List(
      List(Predicate("mother", List("Susan", "Mary"))),
      List(Predicate("mother", List("Susan", "Alice"))),
    )
    val goal =
      List(Predicate("mother", List(new Variable("X"), new Variable("Y"))))
    assertEquals(resolve_pretty(goal, rules), expected)
  }

  test("mother(X, Mary)") {
    val expected = List(
      List(Predicate("mother", List("Susan", "Mary")))
    )
    val goal =
      List(Predicate("mother", List(new Variable("X"), "Mary"))
    )
    assertEquals(resolve_pretty(goal, rules), expected)
  }

  test("fail") {
    val goal = List(Fail("test"))
    try {
      resolve_pretty(goal, rules)
      fail("Expected exception")
    } catch {
      case e: TypeError => assertEquals(e.getMessage, "test")
    }
  }
}
